qrenderdoc/Windows/ShaderViewer.cpp

```diff
@@ -645,6 +645,81 @@ void ShaderViewer::debugShader(const ShaderBindpointMapping *bind, const ShaderR
           }
         }
 
+        // Cache read only and read write resources
+        {
+          m_ReadOnlyResources = m_Ctx.CurPipelineState().GetReadOnlyResources(m_Stage);
+          m_ReadWriteResources = m_Ctx.CurPipelineState().GetReadWriteResources(m_Stage);
+
+          // Mark all resources arrays with more than 8 elements as not dynamically used, then we'll
+          // set the accessed resources as used again
+          for(size_t i = 0; i < m_ReadOnlyResources.size(); ++i)
+          {
+            size_t bindSize = m_ReadOnlyResources[i].resources.size();
+            if(bindSize > 8)
+            {
+              m_ReadOnlyResources[i].dynamicallyUsedCount = 0;
+              for(size_t j = 0; j < bindSize; ++j)
+                m_ReadOnlyResources[i].resources[j].dynamicallyUsed = false;
+            }
+            else
+            {
+              m_ReadOnlyResources[i].dynamicallyUsedCount = (uint32_t)bindSize;
+            }
+          }
+          for(size_t i = 0; i < m_ReadWriteResources.size(); ++i)
+          {
+            size_t bindSize = m_ReadWriteResources[i].resources.size();
+            if(bindSize > 8)
+            {
+              m_ReadWriteResources[i].dynamicallyUsedCount = 0;
+              for(size_t j = 0; j < bindSize; ++j)
+                m_ReadWriteResources[i].resources[j].dynamicallyUsed = false;
+            }
+            else
+            {
+              m_ReadWriteResources[i].dynamicallyUsedCount = (uint32_t)bindSize;
+            }
+          }
+
+          // Walk through the states to find accessed resources
+          size_t numStates = m_States.size();
+          for(size_t s = 0; s < numStates; ++s)
```

Comment:
I don't want to have any passes that iterate over the whole states list up front like this, as in future this list might be lazily populated or dynamic.

Reply:
Refactored to operate on each state and it comes in.